### PR TITLE
feat(autoware_motion_utils): add interpolator

### PR DESCRIPTION
**Pull request by yhisaki**
## Description

An Interpolator class has been added.
The following algorithms have been implemented:

1. AkimaSpline
2. CubicSpline
3. Linear
4. NearestNeighbor
5. Stairstep

For specific implementations and usage, please refer to the [test](common/autoware_motion_utils/test/src/trajectory_container.cpp/test_interpolator.cpp).

## Related links

**Parent Issue:**

- Link

## Notes for reviewers

None.

## Interface changes

None.

## Effects on system behavior

None.

### CODE REVIEW

**Review by Shin-kyoto — CHANGES_REQUESTED** [review]
@yhisaki 

Thank you for sending PR!
Can you make `spell-check-differential` passed by merging https://github.com/tier4/autoware-spell-check-dict/pull/764 ?
Please add the description in the PR above, merge PR and check if the `spell-check-differential` passed.

File: common/autoware_motion_utils/include/autoware/motion_utils/trajectory_container/interpolator/interpolator.hpp [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+   * This method should be overridden by subclasses to return the specific requirement.
+   *
+   * @return The minimum number of required points.
+   */
+  [[nodiscard]] virtual size_t minimum_required_points() const = 0;
+
+  [[nodiscard]] int get_index(const double & s, bool include_end = true) const
+  {
+    if (include_end && s == end()) {
+      return static_cast<int>(axis_.size()) - 2;
```
Comment (Shin-kyoto) [inline common/autoware_motion_utils/include/autoware/motion_utils/trajectory_container/interpolator/interpolator.hpp:169]: I'd like to provide some comments for improvement. Although `int` is frequently used in Autoware, it is actualy an unsafe type. Please use `uint32_t` instaed.
cf. https://matt.sh/howto-c#:~:text=Writing%20code-,Types,-If%20you%20find

Comment (yhisaki) [reply common/autoware_motion_utils/include/autoware/motion_utils/trajectory_container/interpolator/interpolator.hpp:169]: Thank you for your review!! I will fix it.

Comment (yhisaki) [reply common/autoware_motion_utils/include/autoware/motion_utils/trajectory_container/interpolator/interpolator.hpp:169]: I have fixed it. This index is used by eigen, so I use `int32_t`.

File: common/autoware_motion_utils/include/autoware/motion_utils/trajectory_container/interpolator/nearest_neighbor.hpp [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+
+  /**
+   * @brief Compute the interpolated value at the given point.
+   *
+   * @param s The point at which to compute the interpolated value.
+   * @return The interpolated value.
+   */
+  [[nodiscard]] T compute_impl(const double & s) const override
+  {
+    int idx = this->get_index(s);
```
Comment (Shin-kyoto) [inline common/autoware_motion_utils/include/autoware/motion_utils/trajectory_container/interpolator/nearest_neighbor.hpp:59]: @yhisaki 

Can you update all `int` ? 🙏

**Review by Shin-kyoto — CHANGES_REQUESTED** [review]

File: common/autoware_motion_utils/include/autoware/motion_utils/trajectory_container/interpolator/nearest_neighbor.hpp (hunk 2 above)
Comment (yhisaki) [reply common/autoware_motion_utils/include/autoware/motion_utils/trajectory_container/interpolator/nearest_neighbor.hpp:59]: Sorry, I missed it....
I will modify!

Comment (Shin-kyoto) [reply common/autoware_motion_utils/include/autoware/motion_utils/trajectory_container/interpolator/nearest_neighbor.hpp:59]: Thanks a lot!! :+1: :cool:

Comment (yhisaki) [reply common/autoware_motion_utils/include/autoware/motion_utils/trajectory_container/interpolator/nearest_neighbor.hpp]: fixed!

Comment (Shin-kyoto) [reply common/autoware_motion_utils/include/autoware/motion_utils/trajectory_container/interpolator/nearest_neighbor.hpp]: Thanks!
Is it okay to leave `static_cast<int>` as `int` ?

Comment (yhisaki) [reply common/autoware_motion_utils/include/autoware/motion_utils/trajectory_container/interpolator/nearest_neighbor.hpp]: Sorry for making the same mistake over and over again .... I overlooked it.

Comment (Shin-kyoto) [reply common/autoware_motion_utils/include/autoware/motion_utils/trajectory_container/interpolator/nearest_neighbor.hpp]: No, problem!! Thanks :smile:

Comment (yhisaki) [reply common/autoware_motion_utils/include/autoware/motion_utils/trajectory_container/interpolator/nearest_neighbor.hpp]: I think I have fixed all the problems!

File: common/autoware_motion_utils/include/autoware/motion_utils/trajectory_container/interpolator/stairstep.hpp [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+  /**
+   * @brief Compute the interpolated value at the given point.
+   *
+   * @param s The point at which to compute the interpolated value.
+   * @return The interpolated value.
+   */
+  [[nodiscard]] T compute_impl(const double & s) const override
+  {
+    int32_t idx = this->get_index(s, false);
+    return this->values[idx];
```
Comment (Shin-kyoto) [inline common/autoware_motion_utils/include/autoware/motion_utils/trajectory_container/interpolator/stairstep.hpp:60]: @yhisaki 

Please use `.at()` instead.
cf. [TIER IV INTERNAL LINK](https://tier4.atlassian.net/wiki/spaces/CT/pages/3042836632/P+C#Minor)

File: common/autoware_motion_utils/src/trajectory_container/interpolator/akima_spline.cpp [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+#include <cmath>
+#include <vector>
+
+namespace autoware::motion_utils::trajectory_container::interpolator
+{
+
+void AkimaSpline::compute_parameters(
+  const Eigen::Ref<const Eigen::VectorXd> & axis, const Eigen::Ref<const Eigen::VectorXd> & values)
+{
+  int32_t n = static_cast<int32_t>(axis.size());
```
Comment (Shin-kyoto) [inline common/autoware_motion_utils/src/trajectory_container/interpolator/akima_spline.cpp:28]: @yhisaki 

Can you add `const` to variables whose values will not change after initialization?
cf. [TIER IV INTERNAL LINK](https://tier4.atlassian.net/wiki/spaces/CT/pages/3042836632/P+C#Major)

**Review by Shin-kyoto — CHANGES_REQUESTED** [review]

**Review by Shin-kyoto — COMMENTED** [review]
Thanks for updating!

**Review by Shin-kyoto — APPROVED** [review]

File: common/autoware_motion_utils/include/autoware/motion_utils/trajectory_container/interpolator/interpolator_creator.hpp [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+    return *this;
+  }
+
+  [[nodiscard]] std::optional<InterpolatorType> create()
+  {
+    bool success = interpolator_->build(axis_, values_);
+    if (!success) {
+      return std::nullopt;
+    }
+    return interpolator_;
```
Comment (soblin) [inline common/autoware_motion_utils/include/autoware/motion_utils/trajectory_container/interpolator/interpolator_creator.hpp:76]: Is it possible to not to have interpolator_ as member variable, and instead return it here as a value ? I do not like the copy of `interpolator_` to returned value.

Comment (yhisaki) [reply common/autoware_motion_utils/include/autoware/motion_utils/trajectory_container/interpolator/interpolator_creator.hpp]: Thanks! I modified!

**Review by soblin — APPROVED** [review]
LGTM

File: common/autoware_motion_utils/examples/example_interpolator.cpp [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+
+#include <matplotlibcpp17/pyplot.h>
+
+#include <random>
+#include <vector>
+
+int main()
+{
+  pybind11::scoped_interpreter guard{};
+  auto plt = matplotlibcpp17::pyplot::import();
```
Comment (Shin-kyoto) [inline common/autoware_motion_utils/examples/example_interpolator.cpp:27]: @yhisaki 

Can you make PR in https://github.com/tier4/autoware-spell-check-dict to add word?

**Review by Shin-kyoto — CHANGES_REQUESTED** [review]

**Review by takayuki5168 — APPROVED** [review]
LGTM

**Review by Shin-kyoto — APPROVED** [review]
LGTM

**Pull request merged**